.kokoro/nightly/downstream-protobuf-binary-compatibility.sh

```diff
@@ -0,0 +1,107 @@
+monorepo_handwritten_libraries["google-cloud-vertexai"]="com.google.cloud"
+monorepo_handwritten_libraries["google-cloud-resourcemanager"]="com.google.cloud"
+
+# 2. These are the mappings of all the downstream handwritten libraries' artifacts
+declare -A downstream_handwritten_libraries
+downstream_handwritten_libraries["google-cloud"]="com.google.cloud"
+
+# Builds a string output to `artifact_list`. It contains a comma separate list of Maven GAV coordinates. Parses
+# the `versions.txt` file by searching for the matching artifact_id_prefix to get the corresponding version.
+function build_artifact_list() {
+  local -n api_maven_mapping=$1
+  for artifact_id_prefix in "${!api_maven_mapping[@]}"; do
+    group_id="${api_maven_mapping[${artifact_id_prefix}]}"
+
+    # Match all artifacts that start with the artifact_id_prefix to exclude any proto and grpc modules.
+    repo_artifact_list=$(cat "versions.txt" | grep "^${artifact_id_prefix}" || true)
+
+    # Only proceed if there are matching elements
+    if [ -n "${repo_artifact_list}" ]; then
+      # Exclude any matches to BOM artifacts or emulators. The repo artifact list will look like:
+      # "com.google.cloud:google-cloud-accessapproval:2.60.0-SNAPSHOT,com.google.cloud:google-cloud-aiplatform:3.60.0-SNAPSHOT,"
+      repo_artifact_list=$(echo "${repo_artifact_list}" | grep -vE "(bom|emulator|google-cloud-java)" | awk -F: "{\$1=\"${group_id}:\"\$1; \$2=\"\"; print}" OFS=: | sed 's/::/:/' | tr '\n' ',')
+      # Remove the trailing comma after the last entry
+      repo_artifact_list=${repo_artifact_list%,}
+
+      # The first entry added is not separated with a comma. Avoids generating `,{ARTIFACT_LIST}`
+      if [ -z "${artifact_list}" ]; then
+        artifact_list="${repo_artifact_list}"
+      else
+        artifact_list="${artifact_list},${repo_artifact_list}"
+      fi
+    fi
+  done
+}
+
+# cloud-opensource-java contains the Linkage Checker tool
+git clone https://github.com/GoogleCloudPlatform/cloud-opensource-java.git
+pushd cloud-opensource-java
+mvn -B -ntp clean compile -T 1C
+# Linkage Checker tool resides in the /dependencies subfolder
+pushd dependencies
+
+for repo in ${REPOS_UNDER_TEST//,/ }; do # Split on comma
+  # Perform testing on main (with latest changes). Shallow copy as history is not important
+  git clone "https://github.com/googleapis/${repo}.git" --depth=1
+  pushd "${repo}"
+  # Install all repo modules to ~/.m2 (there can be multiple relevant artifacts to test i.e. core, admin, control)
+  mvn -B -ntp install -T 1C -DskipTests -Dclirr.skip -Denforcer.skip
+
+  artifact_list=""
+  if [ "${repo}" == "google-cloud-java" ]; then
+    build_artifact_list monorepo_handwritten_libraries
+  else
+    build_artifact_list downstream_handwritten_libraries
+  fi
+
+  # Linkage Checker /dependencies
+  popd
+
+  echo "Artifact List: ${artifact_list}"
+  # Only run Linkage Checker if the repo has any relevant artifacts to test for
```

Comment:
I guess this is not a realistic scenario, it is in case someone passes in an invalid repo name?

Reply:
Is there a scenario that could cause "no artifacts found"?

Reply:
Currently, no. But adding new downstream tests for repos relies on `versions.txt`'s accuracy. We have few hardcoded values in the script that always assume cloud/ google-cloud/ or com.google.cloud.

Adding something like auth to be tested would lead to `no artifacts found` as it no longer follows `google-cloud` prefix (rather google-auth).

I think it would be harder to figure out the cause if the error message from linkage checker is `Exception in thread "main" java.lang.IllegalArgumentException: Bad artifact coordinates`, rather than `Unable to find any matching artifacts to test in ...`